components/brave_federated/task/model.cc

```diff
@@ -0,0 +1,217 @@
+                                     float loss,
+                                     float accuracy,
+                                     std::vector<Weights> parameters,
+                                     std::map<std::string, double> metrics)
+    : dataset_size(dataset_size),
+      loss(loss),
+      accuracy(accuracy),
+      parameters(parameters),
+      metrics(metrics) {}
+
+PerformanceReport::PerformanceReport(const PerformanceReport& other) = default;
+PerformanceReport::~PerformanceReport() = default;
+
+Model::Model(ModelSpec model_spec)
+    : num_iterations_(model_spec.num_iterations),
+      batch_size_(model_spec.batch_size),
+      learning_rate_(model_spec.learning_rate),
+      threshold_(model_spec.threshold) {
+  const double max_weight = 10.0;
+  const double min_weight = -10.0;
+
+  for (int i = 0; i < model_spec.num_params; i++) {
+    weights_.push_back(base::RandInt(min_weight, max_weight));
+  }
+  bias_ = base::RandInt(min_weight, max_weight);
+}
+
+Model::~Model() = default;
+
+Weights Model::GetWeights() {
+  std::vector<float> copy_of_weights(weights_);
+  return copy_of_weights;
+}
+
+void Model::SetWeights(Weights new_weights) {
+  weights_.assign(new_weights.begin(), new_weights.end());
+}
+
+float Model::GetBias() {
+  return bias_;
+}
+
+void Model::SetBias(float new_bias) {
+  bias_ = new_bias;
+}
+
+size_t Model::ModelSize() {
+  return weights_.size();
+}
+
+std::vector<float> Model::Predict(const DataSet& dataset) {
+  std::vector<float> prediction(dataset.size(), 0.0);
+  for (size_t i = 0; i < dataset.size(); i++) {
+    float z = 0.0;
+    for (size_t j = 0; j < dataset[i].size(); j++) {
+      z += weights_[j] * dataset[i][j];
+    }
+    z += bias_;
+
+    prediction[i] = SigmoidActivation(z);
+  }
+
+  return prediction;
+}
+
+PerformanceReport Model::Train(const DataSet& train_dataset) {
+  auto data_prep_cumulative_duration = base::TimeDelta();
+  auto train_exec_cumulative_duration = base::TimeDelta();
+
+  auto data_start_ts = base::ThreadTicks::Now();
+
+  int features = train_dataset[0].size() - 1;
+  std::vector<float> data_indices(train_dataset.size());
+  for (size_t i = 0; i < train_dataset.size(); i++) {
+    data_indices.push_back(i);
+  }
+
+  Weights d_w(features);
+  std::vector<float> err(batch_size_, 10000);
+  Weights p_w(features);
+  float training_loss = 0.0;
+
+  auto data_end_ts = base::ThreadTicks::Now();
+
+  data_prep_cumulative_duration += base::TimeDelta(data_end_ts - data_start_ts);
+
+  for (int iteration = 0; iteration < num_iterations_; iteration++) {
+    data_start_ts = base::ThreadTicks::Now();
+    base::RandomShuffle(data_indices.begin(), data_indices.end());
+
+    DataSet x(batch_size_, std::vector<float>(features));
+    std::vector<float> y(batch_size_);
+
+    auto exec_start_ts = base::ThreadTicks::Now();
+    for (int i = 0; i < batch_size_; i++) {
+      std::vector<float> point = train_dataset[data_indices[i]];
```

Comment:
Could we be sure that `i` is less than `data_indices.size()`?

Reply:
Good point. Added `DCHECK_LE(batch_size_, train_dataset.size())` and added a check in the task runner that prevents a task from running if there's not enough data.